src/main/java/de/srendi/advancedperipherals/common/util/inventory/InventoryUtil.java

```diff
@@ -31,120 +31,135 @@ private InventoryUtil() {
             if (filter.test(inventoryFrom.getStackInSlot(i))) {
-                ItemStack extracted = inventoryFrom.extractItem(i, amount - transferableAmount, true);
-                if (extracted.isEmpty())
+                ItemStack extracted = inventoryFrom.extractItem(i, required - transferred, true);
+                if (extracted.isEmpty()) {
                     continue;
-                ItemStack inserted;
+                }
+                ItemStack remains;
                 if (toSlot == -1) {
-                    inserted = ItemHandlerHelper.insertItem(inventoryTo, extracted, false);
+                    remains = ItemHandlerHelper.insertItem(inventoryTo, extracted, false);
                 } else {
-                    inserted = inventoryTo.insertItem(toSlot, extracted, false);
+                    remains = inventoryTo.insertItem(toSlot, extracted, false);
+                }
+                int inserted = extracted.getCount() - remains.getCount();
+                if (inserted == 0) {
+                    break;
                 }
-                amount -= inserted.getCount();
-                transferableAmount += inventoryFrom.extractItem(i, extracted.getCount() - inserted.getCount(), false).getCount();
-                if (transferableAmount >= filter.getCount())
+                transferred += inventoryFrom.extractItem(i, inserted, false).getCount();
+                if (transferred >= filter.getCount()) {
                     break;
+                }
             }
         }
-        return transferableAmount;
+        return transferred;
     }
 
     public static int moveFluid(IFluidHandler inventoryFrom, IFluidHandler inventoryTo, FluidFilter filter) {
-        if (inventoryFrom == null) return 0;
+        if (inventoryFrom == null) {
+            return 0;
+        }
 
-        int amount = filter.getCount();
-        int transferableAmount = 0;
+        final int required = filter.getCount();
+        int transferred = 0;
 
         // The logic changes with storage systems since these systems do not have slots
         if (inventoryFrom instanceof IStorageSystemFluidHandler storageSystemHandler) {
             FluidStack extracted = storageSystemHandler.drain(filter, IFluidHandler.FluidAction.SIMULATE);
             int inserted = inventoryTo.fill(extracted, IFluidHandler.FluidAction.EXECUTE);
 
-            transferableAmount += storageSystemHandler.drain(filter.setCount(inserted), IFluidHandler.FluidAction.EXECUTE).getAmount();
+            transferred += storageSystemHandler.drain(filter.setCount(inserted), IFluidHandler.FluidAction.EXECUTE).getAmount();
 
-            return transferableAmount;
+            return transferred;
         }
 
         if (inventoryTo instanceof IStorageSystemFluidHandler storageSystemHandler) {
             if (filter.test(inventoryFrom.getFluidInTank(0))) {
```

Comment:
`getFluidInTank(0)` seems not correct, is that designed for specific case or is it just incomplete yet?

Reply:
Incomplete, not sure why I didn't do that right the first time